api/turing/service/ensembler_service.go

```diff
@@ -0,0 +1,111 @@
+package service
+
+import (
+	"math"
+
+	"github.com/gojek/turing/api/turing/models"
+	"github.com/jinzhu/gorm"
+)
+
+// EnsemblersService is the data access object for the Ensemblers from the db.
+type EnsemblersService interface {
+	// FindByID Find an ensembler matching the given id and options
+	FindByID(id models.ID, options EnsemblersFindByIDOptions) (models.EnsemblerLike, error)
+	// List ensemblers
+	List(options EnsemblersListOptions) (*PaginatedResults, error)
+	// Save the given router to the db. Updates the existing record if already exists
+	Save(ensembler models.EnsemblerLike) (models.EnsemblerLike, error)
+}
+
+type EnsemblersFindByIDOptions struct {
+	ProjectID *models.ID
+}
+
+// EnsemblersListOptions holds query parameters for EnsemblersService.List method
+type EnsemblersListOptions struct {
+	PaginationOptions
+	ProjectID *models.ID `schema:"project_id" validate:"required"`
+}
+
+// NewEnsemblersService creates a new ensemblers service
+func NewEnsemblersService(db *gorm.DB) EnsemblersService {
+	return &ensemblersService{db: db}
+}
+
+type ensemblersService struct {
+	db *gorm.DB
+}
+
+func (service *ensemblersService) FindByID(
+	id models.ID,
+	options EnsemblersFindByIDOptions,
+) (models.EnsemblerLike, error) {
+	var ensembler models.GenericEnsembler
+	query := service.db.Where("id = ?", id)
+
+	if options.ProjectID != nil {
+		query = query.Where("project_id = ?", options.ProjectID)
+	}
+
+	result := query.First(&ensembler)
+	if err := result.Error; err != nil {
+		return nil, err
+	}
+	instance := ensembler.Instance()
+	result = query.Scopes(models.EnsemblerTable(&ensembler)).First(instance)
+	if err := result.Error; err != nil {
+		return nil, err
+	}
+	return instance, nil
+}
+
+func (service *ensemblersService) List(options EnsemblersListOptions) (*PaginatedResults, error) {
+	var results []*models.GenericEnsembler
+	var count int
+	done := make(chan bool, 1)
+
+	query := service.db
+	if options.ProjectID != nil {
+		query = query.Where("project_id = ?", options.ProjectID)
+	}
+
+	go func() {
+		query.Model(&results).Count(&count)
```

Comment:
I'm curious to know how this will turn out. From my understanding, value of results could be in a indeterminate state when line 79 executes as well. I'm not sure if this will have an effect when the value of results changes when line 79 executes.

Reply:
Oh, this is not how gorm works. `&results` here are only used to tell gorm the type of entities (and subsequently, the name of table where it should query the data from). This is effectively the same as `query.Table("ensemblers").Count(&count)` or `query.Model(models.GenericEnsembler{}).Count(&count)`